spring-cloud-data-core/src/main/java/org/springframework/cloud/data/core/StreamDefinition.java

```diff
@@ -16,11 +16,217 @@
 
 package org.springframework.cloud.data.core;
 
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+
+import org.springframework.core.style.ToStringCreator;
+import org.springframework.util.Assert;
+
+/**
+ * Representation of a defined stream. A stream consists of an
+ * ordered list of modules used to process data. Each module
+ * may contain configuration options provided at the time of
+ * stream creation.
+ * <p>
+ * This stream definition does not include any deployment
+ * or runtime configuration for a stream.
+ *
+ * @see ModuleDefinition
+ *
+ * @author Patrick Peralta
+ */
 public class StreamDefinition {
 
-	private String name;
+	/**
+	 * Name of stream.
+	 */
+	private final String name;
 
+	/**
+	 * DSL definition for stream.
+	 */
+	private final String definition;
+
+	/**
+	 * Ordered list of {@link ModuleDefinition}s comprising this stream.
+	 * The source is the first entry and the sink is the last entry.
+	 */
+	private final LinkedList<ModuleDefinition> modules;
+
+	/**
+	 * Construct a {@code StreamDefinition}.
+	 *
+	 * @param name       name of stream
+	 * @param definition DSL definition for stream
+	 * @param modules    ordered list of modules for stream
+	 */
+	private StreamDefinition(String name, String definition, List<ModuleDefinition> modules) {
+		Assert.hasText(name, "name is required");
+		Assert.hasText(definition, "definition is required");
+		Assert.notEmpty(modules, "modules are required");
+		this.name = name;
+		this.definition = definition;
+		this.modules = new LinkedList<>(modules);
+	}
+
+	/**
+	 * Return the name of this stream.
+	 *
+	 * @return stream name
+	 */
 	public String getName() {
 		return name;
 	}
+
+	/**
+	 * Return the DSL definition for this stream.
+	 *
+	 * @return stream definition DSL
+	 */
+	public String getDefinition() {
+		return definition;
+	}
+
+	/**
+	 * Return the ordered list of modules for this stream as a {@link List}.
+	 * This allows for retrieval of modules in the stream by index.
+	 * Modules are maintained in stream flow order (source is first, sink is last).
+	 *
+	 * @return list of module descriptors for this stream definition
+	 */
+	public List<ModuleDefinition> getModuleDefinitions() {
+		return Collections.unmodifiableList(modules);
+	}
+
+	/**
+	 * Return an iterator that indicates the order of module deployments for this
+	 * stream. The modules are returned in reverse order; i.e. the sink is returned
+	 * first followed by the processors in reverse order followed by the
+	 * source.
+	 *
+	 * @return iterator that iterates over the modules in deployment order
+	 */
+	public Iterator<ModuleDefinition> getDeploymentOrderIterator() {
+		return new ReadOnlyIterator<>(modules.descendingIterator());
+	}
+
+	@Override
+	public String toString() {
+		return new ToStringCreator(this)
+				.append("name", this.name)
+				.append(this.definition)
```

Comment:
This is the current format:

`[StreamDefinition@3f9261de name = 'ticktock''time | log']`